### ISSUE TITLE
Wrong result for two simple polygons

### ISSUE THREAD

**Issue by abelramos** [w8r/martinez#110]
These are my two simple polygons:

![Two simple polygons](https://user-images.githubusercontent.com/7587562/71540114-2e3c2600-2914-11ea-84af-921518276a5a.png)

with coordinates:
```
[[115,96], [140,206], [120,210], [125,250], [80,300]]
[[111,228], [129,192], [309,282]]
```

And this is the resulting union:

![Wrong resulting union](https://user-images.githubusercontent.com/7587562/71540115-35fbca80-2914-11ea-98a3-5b7062707284.png)

There are no vertical edges, in fact, no two points share the same x-coordinate. I believe the issue here is with the `otherInOut` property not being correctly computed as these are the sweep events with `otherInOut` set to true after the subdivision step:

![Sweep events with otherInOut set to true](https://user-images.githubusercontent.com/7587562/71540200-61cb8000-2915-11ea-9c7a-f17ba31e116c.png)

I also tried with the C++ code available online from the original paper, and the resulting union is the same.

**Comment by w8r** [comment]
Thanks for the detailed example and the C++ reference, I will try and figure it out

**Comment by jakeNiemiec** [comment]
@abelramos Until this gets fixed, I have found that adding very small (like 0.0005), pseudorandom values to the coordinates will prevent this. **The problem occurs when a point falls exactly on another line.** Our app has grid snapping, I can confirm that this happens all the time.

**Comment by bluenote10** [comment]
I have tested this case with [polybooljs](https://github.com/voidqk/polybooljs) (another implementation of the Martinez algorithm) and there is seems to work:

Input:
![bug_110_input](https://user-images.githubusercontent.com/3620703/71807440-a1a80b00-306b-11ea-9903-6deff1e05695.png)

Output:
![bug_110_output](https://user-images.githubusercontent.com/3620703/71807441-a40a6500-306b-11ea-9214-06f0cb9cbd26.png)

So maybe it helps to have a look where the implementations differ.

**Comment by w8r** [comment]
It's not exactly the same algorithm, but thanks for the reference!

**Comment by bluenote10** [comment]
> I also tried with the C++ code available online from the original paper...

> It's not exactly the same algorithm

Ah interesting, I was wondering about that already. What is this implementation here actually based on? I've found three "original" versions of the Martinez algorithm:
- the 2008 paper, version `cageo141` available as mirror [on GitHub](https://github.com/akavel/martinez-src).
- the 2008 paper, version `cageo144` available from the [Martinez home page (first link)](http://www4.ujaen.es/~fmartin/bool_op.html).
- the 2013 paper, available from the [Martinez home page (last link)](http://www4.ujaen.es/~fmartin/bool_op.html).

**Comment by w8r** [comment]
it's the last one

**Comment by abelramos** [comment]
> @abelramos Until this gets fixed, I have found that adding very small (like 0.0005), pseudorandom values to the coordinates will prevent this. **The problem occurs when a point falls exactly on another line.** Our app has grid snapping, I can confirm that this happens all the time.

Exactly. I can also confirm it is not related to floating-point arithmetic and/or epsilon values, I tested the implementation with a type for rational numbers, and the issue remains.

**Comment by abelramos** [comment]
> I have tested this case with [polybooljs](https://github.com/voidqk/polybooljs)

Yes, I tried with that one too. It works. However, I believe this implementation is faster than the other one, that's why I'm interested in a fix for this one.

**Comment by daef** [comment]
afair [the other one](https://github.com/voidqk/polybooljs) uses only linked lists (instead of a tree), plus solid improvements don't seem to get merged for... [reasons](https://github.com/voidqk/polybooljs/pull/8)

**Comment by w8r** [comment]
I would love to get some help here, I didn't have time to look at it recently

**Comment by bluenote10** [comment]
> afair the other one uses only linked lists (instead of a tree)

Exactly, I came to the same conclusion here after reading the code a bit.

> I would love to get some help here, I didn't have time to look at it recently

I'd like to have a look, but I don't have access to the paper :(.

**Comment by rowanwins** [comment]
@bluenote10 if you send me an email me at rowanwinsemius dot id dot au , then I can supply the paper

**Comment by svenboulanger** [comment]
I'm working on my own implementation of the Martinez paper, and I started there from the original implementation in C++ that they published (I need integer coordinates). Since I was having the same issue: here's what I tracked what was going wrong:

When the status line reaches point (120, 210), the line segments starting at (120,210) are incorrectly inserted between segment (111,228)-(129,192) and (80,300)-(115,96). This causes the segments to have the wrong neighbors.

So what likely needs work, is the segment comparer. I noticed that you're using a comparer somewhat similar to the C++ version so I'm guessing that's also going to apply here. When I made sure the segment comparer ordered them correctly for this one use case, the correct result was being computed.

**Comment by w8r** [comment]
Thank you so much ! I will try that
